File: app/src/main/java/com/storehop/app/ui/settings/HouseholdViewModel.kt
package com.storehop.app.ui.settings

import android.app.Activity
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.storehop.app.billing.BillingManager
import com.storehop.app.billing.Entitlement
import com.storehop.app.billing.EntitlementRepository
import com.storehop.app.billing.isUnlocked
import com.storehop.app.data.entity.HouseholdMember
import com.storehop.app.data.repository.HouseholdRepository
import com.storehop.app.data.repository.InviteCode
import com.storehop.app.data.repository.InviteResult
import com.storehop.app.data.util.HouseholdSessionProvider
import com.storehop.app.data.util.UserSessionProvider
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * v0.7.0 Phase 3b: drives the Settings → Household screen.
 *
 * State combines three sources: current uid, current household id, and the
 * live members list. UI events (generate invite, join with code, leave)
 * delegate to [HouseholdRepository] and surface results through transient
 * fields on [HouseholdUiState] for the UI to render banners + dialogs.
 *
 * Loose error handling: the invite-accept failure cases are typed
 * (NotFound, Expired, AlreadyUsed, Failed) so the UI can render a precise
 * inline error instead of a generic "something went wrong" snackbar.
 */
@HiltViewModel
class HouseholdViewModel @Inject constructor(
    private val repository: HouseholdRepository,
    private val entitlementRepo: EntitlementRepository,
    private val billingManager: BillingManager,
    userSession: UserSessionProvider,
    householdSession: HouseholdSessionProvider,
) : ViewModel() {

    private val _uiEvent = MutableStateFlow<HouseholdUiEvent?>(null)
    val uiEvent: StateFlow<HouseholdUiEvent?> = _uiEvent.asStateFlow()

    /** v0.8: live entitlement state. Generate-Invite button reads this
     *  to decide whether to call [generateInvite] or [launchPurchase]. */
    val entitlement: StateFlow<Entitlement> = entitlementRepo.entitlement

    /** Play-localized formatted price for the upsell button — e.g.
     *  "$7.99" / "€7,99". Null until BillingClient finishes connecting.
     *  Used to render "Unlock for $7.99" instead of hardcoding the
     *  currency on the client. */
    val premiumPrice: StateFlow<String?> = billingManager.productDetails
        .map { it?.oneTimePurchaseOfferDetails?.formattedPrice }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(STOP_TIMEOUT_MS),
            initialValue = null,
        )

    val uiState: StateFlow<HouseholdUiState> = combine(
        userSession.userId,
        householdSession.householdId,
        repository.observeMembers(),
    ) { uid, hid, members ->
        HouseholdUiState(
            currentUid = uid,
            householdId = hid,
            members = members,
            isPersonalHousehold = uid != null && hid == uid,
        )
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(STOP_TIMEOUT_MS),
        initialValue = HouseholdUiState(),
    )

    /**
     * v0.8: launch the Play purchase sheet for the Premium IAP. Wired
     * to the gated "Generate Invite" button when the user is not yet
     * entitled. Activity is required by Play to anchor the sheet's
     * window.
     */
    fun launchPurchase(activity: Activity) {
        billingManager.launchPurchase(activity)
    }

    /**
     * v0.8: combined gate — calls [generateInvite] if entitled, falls
     * back to launching the purchase sheet otherwise. Simplifies the
     * Composable call site (single onClick, VM handles routing).
     */
    fun onGenerateInviteTapped(activity: Activity) {
        if (entitlement.value.isUnlocked) {
            generateInvite()
        } else {
            launchPurchase(activity)
        }
    }

    fun generateInvite() {
        viewModelScope.launch {
            runCatching { repository.generateInvite() }
                .onSuccess { invite -> _uiEvent.value = HouseholdUiEvent.InviteGenerated(invite) }
                .onFailure { e -> _uiEvent.value = HouseholdUiEvent.Failed(e.message ?: "unknown error") }
        }
    }

    fun acceptInvite(rawToken: String) {
        // Normalize: uppercase + strip whitespace + drop visually-ambiguous
        // chars that aren't in the alphabet (in case the user mistyped).
        val token = rawToken.uppercase().filter { it.isLetterOrDigit() }
        if (token.length != INVITE_TOKEN_LENGTH) {
            _uiEvent.value = HouseholdUiEvent.InvalidTokenFormat
            return
        }
        viewModelScope.launch {
            when (val result = repository.acceptInvite(token)) {
                is InviteResult.Success -> _uiEvent.value = HouseholdUiEvent.JoinedHousehold(result.householdId)
                InviteResult.NotFound -> _uiEvent.value = HouseholdUiEvent.InviteNotFound
                InviteResult.Expired -> _uiEvent.value = HouseholdUiEvent.InviteExpired
                InviteResult.AlreadyUsed -> _uiEvent.value = HouseholdUiEvent.InviteAlreadyUsed
                is InviteResult.Failed -> _uiEvent.value =
                    HouseholdUiEvent.Failed(result.cause.message ?: "unknown error")
            }
        }
    }

    fun leaveHousehold() {
        viewModelScope.launch {
            runCatching { repository.leaveHousehold() }
                .onSuccess { _uiEvent.value = HouseholdUiEvent.LeftHousehold }
                .onFailure { e -> _uiEvent.value = HouseholdUiEvent.Failed(e.message ?: "unknown error") }
        }
    }

    fun acknowledgeEvent() {
        _uiEvent.value = null
    }

    companion object {
        private const val INVITE_TOKEN_LENGTH = 8
        private const val STOP_TIMEOUT_MS = 5_000L
    }
}

/** Snapshot of the household screen's combined data. */
data class HouseholdUiState(
    val currentUid: String? = null,
    val householdId: String? = null,
    val members: List<HouseholdMember> = emptyList(),
    val isPersonalHousehold: Boolean = true,
)

/** One-shot outcomes the UI shows as a snackbar or dialog. */
sealed class HouseholdUiEvent {
    data class InviteGenerated(val invite: InviteCode) : HouseholdUiEvent()
    data class JoinedHousehold(val newHouseholdId: String) : HouseholdUiEvent()
    object LeftHousehold : HouseholdUiEvent()
    object InviteNotFound : HouseholdUiEvent()
    object InviteExpired : HouseholdUiEvent()
    object InviteAlreadyUsed : HouseholdUiEvent()
    object InvalidTokenFormat : HouseholdUiEvent()
    data class Failed(val reason: String) : HouseholdUiEvent()
}

File: ios/Storehop/UI/Settings/HouseholdViewModel.swift
import Foundation
import Observation
import SwiftUI

/// v0.7.0 Phase 3b: drives the Settings → Household screen.
///
/// State combines three sources: current uid, current household id, and
/// the live members list. UI events (generate invite, join with code,
/// leave) delegate to `HouseholdRepository` and surface results through
/// transient view-model fields that the SwiftUI view consumes via
/// bindings.
///
/// Mirrors Android's `HouseholdViewModel` 1:1.
@MainActor
@Observable
final class HouseholdViewModel {

    /// The live combined uiState — currentUid, householdId, members list,
    /// and the derived isPersonalHousehold flag.
    var currentUid: String?
    var householdId: String?
    var members: [HouseholdMember] = []
    var isPersonalHousehold: Bool { currentUid != nil && currentUid == householdId }

    /// Form state for the Join card.
    var tokenInput: String = ""
    /// Localized-string key for the join-card inline error. nil = no error.
    var joinErrorKey: String.LocalizationValue?

    /// Pending invite — set to the freshly-generated code so the sheet shows.
    /// Cleared by `dismissPendingInvite()` or after the sheet's Copy / Done.
    var pendingInvite: InviteCode?

    /// Leave-flow alert. The destructive button on the screen flips this
    /// true; the confirm dialog handler flips it back and calls
    /// `confirmLeave()`.
    var showLeaveConfirmation: Bool = false

    /// Failure-alert content. Both invite-network errors and leave errors
    /// surface here as a localized message.
    var failureMessage: String?

    private let repository: any HouseholdRepository
    private let userSession: any UserSessionProvider
    private let householdSession: any HouseholdSessionProvider

    private var uidTask: Task<Void, Never>?
    private var hidTask: Task<Void, Never>?
    private var membersTask: Task<Void, Never>?

    init(
        repository: any HouseholdRepository,
        userSession: any UserSessionProvider,
        householdSession: any HouseholdSessionProvider
    ) {
        self.repository = repository
        self.userSession = userSession
        self.householdSession = householdSession
    }

    func bind() {
        uidTask = Task { [weak self] in
            guard let self else { return }
            for await uid in self.userSession.userIdStream {
                await MainActor.run { self.currentUid = uid }
            }
        }
        hidTask = Task { [weak self] in
            guard let self else { return }
            for await hid in self.householdSession.householdIdStream {
                await MainActor.run { self.householdId = hid }
            }
        }
        membersTask = Task { [weak self] in
            guard let self else { return }
            for await list in self.repository.observeMembers() {
                await MainActor.run { self.members = list }
            }
        }
    }

    func teardown() {
        uidTask?.cancel(); uidTask = nil
        hidTask?.cancel(); hidTask = nil
        membersTask?.cancel(); membersTask = nil
    }

    // MARK: - Invite generate

    func generateInvite() {
        Task { [weak self] in
            guard let self else { return }
            do {
                let code = try await self.repository.generateInvite()
                await MainActor.run { self.pendingInvite = code }
            } catch {
                await MainActor.run {
                    self.failureMessage = error.localizedDescription
                }
            }
        }
    }

    func dismissPendingInvite() {
        pendingInvite = nil
    }

    // MARK: - Invite accept

    /// Client-side 8-char alphanumeric validation, then dispatch to repo.
    func acceptInvite() {
        let normalized = tokenInput
            .uppercased()
            .filter { $0.isLetter || $0.isNumber }
        guard normalized.count == Self.inviteTokenLength else {
            joinErrorKey = "household_error_invalid_token"
            return
        }
        joinErrorKey = nil
        Task { [weak self] in
            guard let self else { return }
            switch await self.repository.acceptInvite(token: normalized) {
            case .success:
                await MainActor.run {
                    self.tokenInput = ""
                    self.joinErrorKey = nil
                }
            case .notFound:
                await MainActor.run { self.joinErrorKey = "household_error_invite_not_found" }
            case .expired:
                await MainActor.run { self.joinErrorKey = "household_error_invite_expired" }
            case .alreadyUsed:
                await MainActor.run { self.joinErrorKey = "household_error_invite_used" }
            case .failed(let reason):
                await MainActor.run { self.failureMessage = reason }
            }
        }
    }

    // MARK: - Leave flow

    func requestLeaveConfirmation() { showLeaveConfirmation = true }
    func cancelLeave() { showLeaveConfirmation = false }

    func confirmLeave() {
        showLeaveConfirmation = false
        Task { [weak self] in
            guard let self else { return }
            do {
                try await self.repository.leaveHousehold()
            } catch {
                await MainActor.run {
                    self.failureMessage = error.localizedDescription
                }
            }
        }
    }

    func acknowledgeFailure() { failureMessage = nil }

    // MARK: - SwiftUI binding helpers

    var pendingInviteBinding: Binding<InviteCode?> {
        Binding(
            get: { self.pendingInvite },
            set: { newValue in
                if newValue == nil { self.pendingInvite = nil }
            }
        )
    }

    var leaveConfirmationBinding: Binding<Bool> {
        Binding(
            get: { self.showLeaveConfirmation },
            set: { self.showLeaveConfirmation = $0 }
        )
    }

    var failedEventBinding: Binding<Bool> {
        Binding(
            get: { self.failureMessage != nil },
            set: { showing in
                if !showing { self.failureMessage = nil }
            }
        )
    }

    private static let inviteTokenLength = 8
}

extension InviteCode: Identifiable {
    /// SwiftUI `sheet(item:)` requires Identifiable. The token already is
    /// unique per invite, so reuse it as the identity.
    var id: String { token }
}
